flask_appbuilder/tests/test_api.py

```diff
@@ -3100,3 +3100,350 @@ def test_before_request(self):
         uri = "api/v1/model1beforerequest/update"
         rv = self.auth_client_put(client, token, uri, {})
         self.assertEqual(rv.status_code, 404)
+
+
+class UserAPITestCase(FABTestCase):
+    def setUp(self):
+        from flask import Flask
+        from flask_appbuilder import AppBuilder
+
+        self.app = Flask(__name__)
+        self.basedir = os.path.abspath(os.path.dirname(__file__))
+        self.app.config.from_object("flask_appbuilder.tests.config_api")
+        self.app.config["ENABLE_USER_CRUD_API"] = True
+        self.db = SQLA(self.app)
+        self.appbuilder = AppBuilder(self.app, self.db.session)
+
+        # TODO: this heinous hack is to avoid using stale db session leaking from
+        # RolePermissionAPITestCase
+        # don't know why all baseviews in Appbuilder are attached to stale session,
+        # causing error when adding a new user which reads roles from this session and
+        # datamodel uses stale session to add it.
+        for b in self.appbuilder.baseviews:
+            if hasattr(b, "datamodel") and b.datamodel.session is not None:
+                b.datamodel.session = self.db.session
+
+    def tearDown(self):
+        self.appbuilder.get_session.close()
+        engine = self.db.session.get_bind(mapper=None, clause=None)
+        engine.dispose()
+
+    def test_user_api(self):
+        """REST Api: Test user apis
+        """
+        client = self.app.test_client()
+        token = self.login(client, USERNAME_ADMIN, PASSWORD_ADMIN)
+
+        uri = "api/v1/user/"
+        rv = self.auth_client_get(client, token, uri)
+        self.assertEqual(rv.status_code, 200)
+
+        uri = "api/v1/user/1"
+        rv = self.auth_client_get(client, token, uri)
+        self.assertEqual(rv.status_code, 200)
+
+        uri = "api/v1/user/"
+        create_user_payload = {
+            "active": True,
+            "email": "fab@usearapiteaasdst.com",
+            "first_name": "fab",
+            "last_name": "admin",
+            "password": "password",
+            "roles": [1],
+            "username": "fasab_usear_api_test",
+        }
+        rv = self.auth_client_post(client, token, uri, create_user_payload)
+        add_user_response = json.loads(rv.data)
+        self.assertEqual(rv.status_code, 201)
+        assert "id" and "result" in add_user_response
+        self.assertEqual(create_user_payload, add_user_response["result"])
+
+        uri = f"api/v1/user/{add_user_response['id']}"
```

Comment:
let's test the crud on separate tests. One to get list, one got get item, one for non existent get item, one for create user, a couple for create user edge cases (no roles, invalid password etc), one for update and update edge cases/error conditions

Reply:
done.